### ISSUE TITLE
Amplify Datastore clear not working in iOS

### ISSUE THREAD

**Issue by djorgji** [aws-amplify/amplify-flutter#1470]
### Description

Data not clearing when running `clear()` on Datastore. The authenticator properly shows the login screen again on logout, however login again, if a different users still shows old users data, based on sync expressions the new users data is the only data syncing the old users data stays stale.

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [ ] Auth
- [ ] Authenticator
- [X] DataStore
- [ ] Storage

### Steps to Reproduce

Running on user button click:

```
  Future<void> signOut() async {
    await Amplify.DataStore.stop();
    await Amplify.DataStore.clear();
    await Amplify.Auth.signOut();
  }
```

I have tried reordering these a couple of times or running them with `.then()` instead nothing has worked properly on iOS, it works on android. OR (which I prefer because it is a much more snappy experience for the user, as stop/clear take too long to execute)
```
  Future<void> signOut() async {
    await Amplify.Auth.signOut();
    await Amplify.DataStore.stop();
    await Amplify.DataStore.clear();
  }
```

### Platforms

- [X] iOS
- [ ] Android

### Environment

```bash
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.10.3, on macOS 12.2.1 21D62 darwin-arm, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
[✓] Xcode - develop for iOS and macOS (Xcode 13.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.1)
[✓] VS Code (version 1.65.2)
[✓] Connected device (3 available)
[✓] HTTP Host Availability

• No issues found!
```

### Dependencies

```bash
Dart SDK 2.16.1
Flutter SDK 2.10.3
chorus_devices 1.0.0

dependencies:
- adaptive_navigation 0.0.5 [flutter adaptive_breakpoints]
- amplify_api 0.4.1 [amplify_api_plugin_interface amplify_core collection flutter meta plugin_platform_interface]
- amplify_auth_cognito 0.4.1 [flutter amplify_auth_plugin_interface amplify_core collection plugin_platform_interface]
- amplify_authenticator 0.1.0 [amplify_auth_cognito amplify_auth_plugin_interface amplify_core amplify_flutter collection flutter flutter_localizations intl]
- amplify_datastore 0.4.1 [flutter amplify_datastore_plugin_interface amplify_core plugin_platform_interface meta collection async]
- amplify_flutter 0.4.1 [amplify_analytics_plugin_interface amplify_api_plugin_interface amplify_auth_plugin_interface amplify_core amplify_datastore_plugin_interface amplify_storage_plugin_interface collection flutter json_annotation meta plugin_platform_interface]
- async 2.8.2 [collection meta]
- badges 2.0.2 [flutter]
- cbor 5.0.1 [collection typed_data hex convert ieee754 meta]
- cupertino_icons 1.0.4
- firebase_core 1.13.1 [firebase_core_platform_interface firebase_core_web flutter meta]
- firebase_crashlytics 2.5.3 [firebase_core firebase_core_platform_interface firebase_crashlytics_platform_interface flutter stack_trace]
- firebase_messaging 11.2.11 [firebase_core firebase_core_platform_interface firebase_messaging_platform_interface firebase_messaging_web flutter meta]
- fl_chart 0.46.0 [flutter equatable]
- flutter 0.0.0 [characters collection material_color_utilities meta typed_data vector_math sky_engine]
- flutter_local_notifications 9.4.0 [clock flutter flutter_local_notifications_linux flutter_local_notifications_platform_interface timezone]
- flutter_localizations 0.0.0 [flutter intl characters clock collection material_color_utilities meta path typed_data vector_math]
- flutter_reactive_ble 5.0.2 [collection flutter functional_data meta reactive_ble_mobile reactive_ble_platform_interface]
- flutter_spinkit 5.1.0 [flutter]
- fluttertoast 8.0.9 [flutter flutter_web_plugins]
- http 0.13.4 [async http_parser meta path]
- intl 0.17.0 [clock path]
- path_provider 2.0.9 [flutter path_provider_android path_provider_ios path_provider_linux path_provider_macos path_provider_platform_interface path_provider_windows]
- path_to_regexp 0.4.0
- permission_handler 9.2.0 [flutter meta permission_handler_android permission_handler_apple permission_handler_windows permission_handler_platform_interface]
- provider 6.0.2 [collection flutter nested]
- quiver 3.0.1+1 [matcher]
- shared_preferences 2.0.13 [flutter shared_preferences_android shared_preferences_ios shared_preferences_linux shared_preferences_macos shared_preferences_platform_interface shared_preferences_web shared_preferences_windows]
- url_launcher 6.0.20 [flutter url_launcher_android url_launcher_ios url_launcher_linux url_launcher_macos url_launcher_platform_interface url_launcher_web url_launcher_windows]

transitive dependencies:
- adaptive_breakpoints 0.0.5 [flutter]
- amplify_analytics_plugin_interface 0.4.1 [amplify_core flutter meta]
- amplify_api_plugin_interface 0.4.1 [amplify_core collection flutter json_annotation meta]
- amplify_auth_plugin_interface 0.4.1 [flutter meta amplify_core]
- amplify_core 0.4.1 [flutter plugin_platform_interface collection date_time_format meta uuid]
- amplify_datastore_plugin_interface 0.4.1 [flutter meta collection amplify_core]
- amplify_storage_plugin_interface 0.4.1 [flutter meta amplify_core]
- args 2.3.0
- characters 1.2.0
- charcode 1.3.1
- clock 1.1.0
- collection 1.15.0
- convert 3.0.1 [typed_data]
- crypto 3.0.1 [collection typed_data]
- date_time_format 2.0.1
- dbus 0.7.2 [args ffi meta xml]
- equatable 2.0.3 [collection meta]
- ffi 1.1.2
- file 6.1.2 [meta path]
- firebase_core_platform_interface 4.2.5 [collection flutter meta plugin_platform_interface]
- firebase_core_web 1.6.1 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- firebase_crashlytics_platform_interface 3.2.1 [collection firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_platform_interface 3.2.1 [firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_web 2.2.9 [firebase_core firebase_core_web firebase_messaging_platform_interface flutter flutter_web_plugins js meta]
- fixnum 1.0.0
- flutter_local_notifications_linux 0.4.2 [flutter flutter_local_notifications_platform_interface dbus path xdg_directories]
- flutter_local_notifications_platform_interface 5.0.0 [flutter plugin_platform_interface]
- flutter_web_plugins 0.0.0 [flutter js characters collection material_color_utilities meta typed_data vector_math]
- functional_data 1.0.0 [meta collection]
- hex 0.2.0
- http_parser 4.0.0 [charcode collection source_span string_scanner typed_data]
- ieee754 1.0.3 [meta]
- js 0.6.3
- json_annotation 4.4.0 [meta]
- matcher 0.12.11 [stack_trace]
- material_color_utilities 0.1.3
- meta 1.7.0
- nested 1.0.0 [flutter]
- path 1.8.0
- path_provider_android 2.0.12 [flutter path_provider_platform_interface]
- path_provider_ios 2.0.8 [flutter path_provider_platform_interface]
- path_provider_linux 2.1.5 [ffi flutter path path_provider_platform_interface xdg_directories]
- path_provider_macos 2.0.5 [flutter path_provider_platform_interface]
- path_provider_platform_interface 2.0.3 [flutter platform plugin_platform_interface]
- path_provider_windows 2.0.5 [ffi flutter path path_provider_platform_interface win32]
- permission_handler_android 9.0.2+1 [flutter permission_handler_platform_interface]
- permission_handler_apple 9.0.3 [flutter permission_handler_platform_interface]
- permission_handler_platform_interface 3.7.0 [flutter meta plugin_platform_interface]
- permission_handler_windows 0.1.0 [flutter permission_handler_platform_interface]
- petitparser 4.4.0 [meta]
- platform 3.1.0
- plugin_platform_interface 2.1.2 [meta]
- process 4.2.4 [file path platform]
- protobuf 2.0.1 [fixnum collection]
- reactive_ble_mobile 5.0.2 [flutter protobuf reactive_ble_platform_interface]
- reactive_ble_platform_interface 5.0.2 [collection flutter functional_data meta plugin_platform_interface]
- shared_preferences_android 2.0.11 [flutter shared_preferences_platform_interface]
- shared_preferences_ios 2.1.0 [flutter shared_preferences_platform_interface]
- shared_preferences_linux 2.1.0 [file flutter path path_provider_linux path_provider_platform_interface shared_preferences_platform_interface]
- shared_preferences_macos 2.0.3 [flutter shared_preferences_platform_interface]
- shared_preferences_platform_interface 2.0.0 [flutter]
- shared_preferences_web 2.0.3 [flutter flutter_web_plugins shared_preferences_platform_interface]
- shared_preferences_windows 2.1.0 [file flutter path path_provider_platform_interface path_provider_windows shared_preferences_platform_interface]
- sky_engine 0.0.99
- source_span 1.8.1 [collection path term_glyph]
- stack_trace 1.10.0 [path]
- string_scanner 1.1.0 [charcode source_span]
- term_glyph 1.2.0
- timezone 0.8.0 [path]
- typed_data 1.3.0 [collection]
- url_launcher_android 6.0.15 [flutter url_launcher_platform_interface]
- url_launcher_ios 6.0.15 [flutter url_launcher_platform_interface]
- url_launcher_linux 3.0.0 [flutter url_launcher_platform_interface]
- url_launcher_macos 3.0.0 [flutter url_launcher_platform_interface]
- url_launcher_platform_interface 2.0.5 [flutter plugin_platform_interface]
- url_launcher_web 2.0.9 [flutter flutter_web_plugins url_launcher_platform_interface]
- url_launcher_windows 3.0.0 [flutter url_launcher_platform_interface]
- uuid 3.0.6 [crypto]
- vector_math 2.1.1
- win32 2.4.2 [ffi]
- xdg_directories 0.2.0+1 [meta path process]
- xml 5.3.1 [collection meta petitparser]
```

### Device

iPhone 7 Plus, iPhone 12, iPhone 13, One Plus N200

### OS

iOS 15.0.2, Android 11

### CLI Version

7.6.22

### Additional Context

Uninstalling and reinstalling the app is the only workaround I have been able to find.

**Comment by HuiSF** [comment]
Hello @djorgji thank for opening this issue.

**TL;DR**

If you are going to invoke `Amplify.DataStore.clear()`, you don't need to invoke `Amplify.DataStore.stop()` before that.

**Context**

There is a known issue that `Amplify.DataStore.clear` doesn't actually work when the underlying DataStore sync engine is stopped (could be stopped by invoking `Amplify.DataStore.stop()`) in amplify-ios. Amplify iOS maintainers have implemented a fix, and amplify-flutter is going to integrate the fix and release with version `0.4.2`.

**Comment by djorgji** [comment]
Workaround worked, thanks!

**Comment by HuiSF** [comment]
With the latest release of [0.4.2](https://github.com/aws-amplify/amplify-flutter/releases/tag/v0.4.2), consecutively calling `DataStore.stop()`, `DataStore.clear()` in iOS, the clear functionality should work as expected.

But still, no need to call `DataStore.stop()`, `DataStore.clear()` is going to be called.
